ext/socket/lib/socket.rb

```diff
@@ -623,45 +684,131 @@ def accept_nonblock(exception: true)
+    end
 
     local_addr_list = nil
-    if local_host != nil || local_port != nil
+    if local_host && local_port
       local_addr_list = Addrinfo.getaddrinfo(local_host, local_port, nil, :STREAM, nil)
+      attempt_v6 = local_addr_list.any? {|local_ai| local_ai.afamily == Socket::AF_INET6 }
+      attempt_v4 = local_addr_list.any? {|local_ai| local_ai.afamily == Socket::AF_INET }
+    end
+
+    queue = Queue.new
+
+    if attempt_v6
+      getaddrinfo_v6_th = start_getaddrinfo_v6(host, port, queue, resolv_timeout)
+    end
+
+    if attempt_v4
+      getaddrinfo_v4_th = start_getaddrinfo_v4(host, port, queue, resolv_timeout)
+    end
+
+    if connect_timeout
+      timeout_th = Thread.new do
+        sleep(connect_timeout)
+        queue.close()
+      end
     end
 
-    Addrinfo.foreach(host, port, nil, :STREAM, timeout: resolv_timeout) {|ai|
-      if local_addr_list
-        local_addr = local_addr_list.find {|local_ai| local_ai.afamily == ai.afamily }
-        next unless local_addr
+    socket_list = []
+    getaddrinfo_v6_done = attempt_v6 ? false : true
+    getaddrinfo_v4_done = attempt_v4 ? false : true
+    while ai = queue.pop do
+      break if queue.closed? # timed out
+      case ai
+      when :getaddrinfo_v6_done
+        getaddrinfo_v6_done = true
+      when :getaddrinfo_v4_done
+        getaddrinfo_v4_done = true
       else
         local_addr = nil
+        if local_addr_list
+          local_addr = local_addr_list.find {|local_ai| local_ai.afamily == ai.afamily }
+          next unless local_addr
+        end
+        begin
+          sock = Socket.new(ai.pfamily, ai.socktype, ai.protocol)
+          sock.ipv6only! if ai.ipv6?
+          sock.bind(local_addr) if local_addr
+          case sock.connect_nonblock(ai, exception: false)
+          when 0
+            ret = sock
+            break
+          when :wait_writable
+            socket_list.push(sock)
+            # 250ms is the recommended value for the connection attempt delay in RFC8305
+            _, writable_sockets, = IO.select(nil, socket_list, nil, 0.25)
+            ret, error = find_connected_socket(writable_sockets)
+            last_error = error if error
+            break if ret
+          end
+        rescue SystemCallError => ex
+          last_error = ex
+        end
+      end
+      if getaddrinfo_v6_done && getaddrinfo_v4_done && queue.empty?
+        break # all connection attempts has been made
       end
+    end
+
+    # wait for sockets to be connected
+    if !ret && !socket_list.empty?
       begin
-        sock = local_addr ?
-          ai.connect_from(local_addr, timeout: connect_timeout) :
-          ai.connect(timeout: connect_timeout)
-      rescue SystemCallError
-        last_error = $!
-        next
+        if connect_timeout
+          if queue.closed? # timed out
+            timeout = 0 # returns immediately
+          else
+            # wait until connect_timeout
+            elapsed = Time.now - start_time
```

Comment:
Prefer `Process.clock_gettime(Process::CLOCK_MONOTONIC)` instead of `Time.now` to measure elapsed time.

Reply:
Good point. I will use the monotonic clock.